chain/chain/src/resharding/resharding_actor.rs

```diff
@@ -17,51 +16,42 @@ pub struct ReshardingActor {
 
 impl messaging::Actor for ReshardingActor {}
 
-impl Handler<FlatStorageSplitShardRequest> for ReshardingActor {
-    fn handle(&mut self, msg: FlatStorageSplitShardRequest) {
-        match msg.resharder.split_shard_task() {
-            FlatStorageReshardingTaskStatus::Successful { .. } => {
+impl HandlerWithContext<FlatStorageSplitShardRequest> for ReshardingActor {
+    fn handle(
+        &mut self,
+        msg: FlatStorageSplitShardRequest,
+        ctx: &mut dyn DelayedActionRunner<Self>,
+    ) {
+        self.handle_flat_storage_split_shard(msg.resharder, ctx);
+    }
+}
+
+impl Handler<FlatStorageShardCatchupRequest> for ReshardingActor {
+    fn handle(&mut self, msg: FlatStorageShardCatchupRequest) {
+        match msg.resharder.shard_catchup_task(
+            msg.shard_uid,
+            msg.flat_head_block_hash,
+            &self.chain_store,
+        ) {
+            FlatStorageReshardingTaskResult::Successful { .. } => {
                 // All good.
             }
-            FlatStorageReshardingTaskStatus::Failed => {
-                panic!("impossible to recover from a flat storage split shard failure!")
+            FlatStorageReshardingTaskResult::Failed => {
+                panic!("impossible to recover from a flat storage shard catchup failure!")
             }
-            FlatStorageReshardingTaskStatus::Cancelled => {
-                // The task has been cancelled. Nothing else to do.
+            FlatStorageReshardingTaskResult::Cancelled => {
+                panic!("shard catchup task should never be cancelled!")
             }
-            FlatStorageReshardingTaskStatus::Postponed => {
-                // The task has been postponed for later execution. Nothing to do.
+            FlatStorageReshardingTaskResult::Postponed => {
+                panic!("shard catchup task should never be postponed!")
             }
```

Comment:
Is it possible for those cases to be triggered? If not perhaps you can introduce a smaller type for the catchup result?

Reply:
Indeed, I'll introduce another type to avoid these 'impossible' cases